### PR TITLE
Parse mapping file passed to makemsix

### PR DESCRIPTION
**Pull request by jyvenugo**
- Created a MappingFileParser.cpp file which parses the heading section in the mapping file followed by each of 
   its mappings.
- I have only stored mappings for the following two sections, 'Files' and 'ExternalPackages'. I have not stored 
   mappings for 'Keys' and 'ResourceMetadata. 
- Skipped the section if the section listed in the mapping file is not one of the four known sections. 
- I checked for the following parsing errors in the mapping file,
   Empty lines, Missing token quotes, too less or too few tokens on a line, missing whitespace between two 
   tokens on the same line, duplicate sections, missing end bracket in section heading, missing Files section in 
   mapping file, lines not starting with [ or ".
- I stored these mappings in maps(FileList and ExternalPackages) in the format std::map<outputFile, InputFile> 
   and passed these values to the ParseBundlePayloadFromMappingFile method in AppxBundleWriter.cpp from msix.cpp.
- Creating the flat bundle from these values passed in from the mapping file. 
- Implemented AddExternalPackageReference to add the ExternalPackages to the bundle.
Verified by running makemsix.exe to create the flat bundle(for eg., makemsix.exe bundle -f [path to ..MappingFile.txt] -p [path to .. output result.msixbundle] -fb -bv 1.0.1.2096), Verified that the flat bundle gets create and I am able to install it as well.

### CODE REVIEW

**Review by wcheng-msft — APPROVED** [review]
:shipit:

**Review by wcheng-msft — APPROVED** [review]

File: src/msix/msix.cpp [hunk 1]
```diff
@@ -348,18 +348,6 @@ MSIX_API HRESULT STDMETHODCALLTYPE PackBundle(
     }
     //TODO:: Error if directoryPath is a file
```
Comment (msftrubengu) [inline src/msix/msix.cpp:349]: >//TODO:: Error if directoryPath is a file [](start = 4, length = 41)

Remove

File: src/msix/msix.cpp [hunk 2]
```diff
@@ -380,7 +368,24 @@ MSIX_API HRESULT STDMETHODCALLTYPE PackBundle(
...
+        MSIX::MappingFileParser mappingFileParser;
+        mappingFileParser.ParseMappingFile(mappingFile);
+        if(!mappingFileParser.IsSectionFound(MSIX::SectionID::FilesSection))
+        {
+            std::ostringstream errorBuilder;
+            errorBuilder << "The mapping file " << mappingFile << " is missing a [Files] section.";
+            ThrowErrorAndLog(MSIX::Error::BadFormat, errorBuilder.str().c_str());
+        }
+        bundleWriter4.As<IBundleWriter>()->ProcessBundlePayloadFromMappingFile(mappingFileParser.GetFileList(), mappingFileParser.GetExternalPackagesList(), flatBundle);
+    }
```
Comment (msftrubengu) [inline src/msix/msix.cpp:387]: >} [](start = 4, length = 1)

i think we should fail with invalid args in the else

File: src/inc/internal/MappingFileParser.hpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        {"ExternalPackages", ExternalPackagesSection}};
+
+    class MappingFileParser final
+    {
+    public:
+        MappingFileParser();
+        void ParseMappingFile(std::string mappingFile);
+        bool IsSectionFound(SectionID sectionId);
+        std::map<std::string, std::string> GetFileList() { return this->list; }
+        std::map<std::string, std::string> GetExternalPackagesList() { return this->externalPackagesList; }
```
Comment (msftrubengu) [inline src/inc/internal/MappingFileParser.hpp:42]: As a general practice, add const at the end of the the method declaration on this type of methods. 
eg. std::map<std::string, std::string> GetFileList() const { return this->list; }

File: src/inc/internal/BundleWriterHelper.hpp [hunk 4]
```diff
@@ -73,7 +71,9 @@ namespace MSIX {
         std::string mainPackageName;
         std::string mainPackagePublisher;
         std::uint64_t bundleVersion;
+        //std::uint64_t payloadPackageMinWinTargetDeviceFamilyVersionForWindows = UINT64_MAX;
```
Comment (msftrubengu) [inline src/inc/internal/BundleWriterHelper.hpp:74]: >//std::uint64_t payloadPackageMinWinTargetDeviceFamilyVersionForWindows = UINT64_MAX; [](start = 8, length = 85)

remove

File: src/inc/internal/MappingFileParser.hpp (hunk 3 above)
Comment (msftrubengu) [reply src/inc/internal/MappingFileParser.hpp:42]: you can also return a const reference and make it more efficient. This is only true if the caller wont modify the maps
const std::map<std::string, std::string>& GetFileList() const { return this->list; }

---
In reply to: [532771453](https://github.com/microsoft/msix-packaging/pull/395#discussion_r532771453) [](ancestors = 532771453)

File: src/inc/internal/AppxManifestObject.hpp [hunk 5]
```diff
@@ -60,6 +60,22 @@ class IAppxManifestQualifiedResourceInternal : public IUnknown
...
+#else
+#include "Unknwn.h"
+#include "Objidl.h"
+class IAppxManifestMainPackageDependencyInternal : public IUnknown
+#endif
+{
+public:
+    virtual const std::string& GetName() = 0;
+    virtual const std::string& GetPublisher() = 0;
+    virtual const std::string& GetPackageFamilyName() = 0;
```
Comment (msftrubengu) [inline src/inc/internal/AppxManifestObject.hpp:75]: const std::string& GetName() const

File: src/msix/pack/MappingFileParser.cpp [hunk 6]
```diff
@@ -0,0 +1,283 @@
+#include "MappingFileParser.hpp"
+#include <iostream>
+#include <fstream>
+#include <map>
+
+namespace MSIX {
+
+    const std::string WHITESPACE = " \t";
```
Comment (msftrubengu) [inline src/msix/pack/MappingFileParser.cpp:8]: >WHITESPACE [](start = 22, length = 10)

TAB?

File: src/msix/pack/BundleWriterHelper.cpp [hunk 7]
```diff
@@ -211,13 +165,117 @@ namespace MSIX {
...
 
-    void BundleWriterHelper::EndBundleManifest()
+    void BundleWriterHelper::AddExternalPackageReferenceFromManifest(std::string fileName, IAppxManifestReader* manifestReader,
+        bool isDefaultApplicablePackage)
     {
+        ComPtr<IAppxManifestReader5> manifestReader5;
+        ThrowHrIfFailed(manifestReader->QueryInterface(UuidOfImpl<IAppxManifestReader5>::iid, reinterpret_cast<void**>(&manifestReader5)));
+        ComPtr<IAppxManifestMainPackageDependenciesEnumerator> mainPackageDependencies;
+        ThrowHrIfFailed(manifestReader5->GetMainPackageDependencies(&mainPackageDependencies));
+        BOOL hasMoreMainPackageDependencies = FALSE;
```
Comment (msftrubengu) [inline src/msix/pack/BundleWriterHelper.cpp:175]: >BOOL hasMoreMainPackageDependencies = FALSE; [](start = 8, length = 44)

use bool

File: src/msix/pack/BundleValidationHelper.cpp [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
+    void BundleValidationHelper::ValidateTargetDeviceFamiliesFromManifestPackageId(IAppxManifestPackageIdInternal* packageId, IAppxManifestTargetDeviceFamiliesEnumerator* tdfs, std::string fileName)
+    {
+        std::string arch = packageId->GetArchitecture();
+        if (arch == "neutral")
+        {
+            this->numberOfNeutralAppPackages++;
+        }
+
+        BOOL hasCurrent = FALSE;
```
Comment (msftrubengu) [inline src/msix/pack/BundleValidationHelper.cpp:19]: >BOOL hasCurrent = FALSE; [](start = 8, length = 24)

same

File: src/msix/pack/BundleValidationHelper.cpp [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    const std::string Win10DeviceFamilyNameStartsWith = "Windows.";
+    const std::string OSVersionRS5 = "10.0.17763.0";
+
+    BundleValidationHelper::BundleValidationHelper() 
+    {}
+
+    void BundleValidationHelper::ValidateTargetDeviceFamiliesFromManifestPackageId(IAppxManifestPackageIdInternal* packageId, IAppxManifestTargetDeviceFamiliesEnumerator* tdfs, std::string fileName)
+    {
+        std::string arch = packageId->GetArchitecture();
+        if (arch == "neutral")
```
Comment (msftrubengu) [inline src/msix/pack/BundleValidationHelper.cpp:14]: >"neutral" [](start = 20, length = 9)

make this a const std::string as Win10DeviceFamilyNameStartsWith and use it i this method

File: src/msix/pack/AppxBundleWriter.cpp [hunk 10]
```diff
@@ -72,6 +72,52 @@ namespace MSIX {
...
+    void AppxBundleWriter::ProcessBundlePayloadFromMappingFile(std::map<std::string, std::string> fileList, std::map<std::string, std::string> externalPackagesList, bool flatBundle)
+    {
+        ThrowErrorIf(Error::InvalidState, m_state != WriterState::Open, "Invalid package writer state");
+        auto failState = MSIX::scope_exit([this]
+            {
+                this->m_state = WriterState::Failed;
+            });
+
+        std::map<std::string, std::string>::iterator fileListIterator;
+        for (fileListIterator = fileList.begin(); fileListIterator != fileList.end(); fileListIterator++)
```
Comment (msftrubengu) [inline src/msix/pack/AppxBundleWriter.cpp:84]: You could also do
for (const auto& file : fileList)

**Review by msftrubengu — APPROVED** [review]
:shipit:

File: src/msix/msix.cpp (hunk 1 above)
Comment (jyvenugo) [reply src/msix/msix.cpp:358]: I am keeping this for now, to remember the TODO. I need to add this portion to the code.

File: src/inc/internal/BundleWriterHelper.hpp (hunk 4 above)
Comment (jyvenugo) [reply src/inc/internal/BundleWriterHelper.hpp:74]: Removed

File: src/msix/pack/BundleValidationHelper.cpp (hunk 9 above)
Comment (jyvenugo) [reply src/msix/pack/BundleValidationHelper.cpp:14]: Updated.

File: src/msix/pack/MappingFileParser.cpp (hunk 6 above)
Comment (jyvenugo) [reply src/msix/pack/MappingFileParser.cpp:8]: As per other implementations, I see whitespace and tab as the delimiter.

File: src/msix/msix.cpp (hunk 2 above)
Comment (jyvenugo) [reply src/msix/msix.cpp:455]: I modified the code different from how I had it before to handle other cases as well.

File: src/inc/internal/MappingFileParser.hpp (hunk 3 above)
Comment (jyvenugo) [reply src/inc/internal/MappingFileParser.hpp]: Updated to reflect this.

**Pull request merged**